src/lib/mac/hmac/hmac.cpp

```diff
@@ -54,27 +55,51 @@ void HMAC::key_schedule(const uint8_t key[], size_t length)
    m_hash->clear();
 
    m_ikey.resize(m_hash_block_size);
-   set_mem(m_ikey.data(), m_hash_block_size, ipad);
-
    m_okey.resize(m_hash_block_size);
-   set_mem(m_okey.data(), m_hash_block_size, opad);
+
+   clear_mem(m_ikey.data(), m_ikey.size());
+   clear_mem(m_okey.data(), m_okey.size());
+
+   /*
+   * Sometimes the HMAC key length itself is sensitive, as with PBKDF2 where it
+   * reveals the length of the passphrase. Make some attempt to hide this to
+   * side channels. Clearly if the secret is longer than the block size then the
+   * branch to hash first reveals that. In addition, counting the number of
+   * compression functions executed reveals the size at the granularity of the
+   * hash function's block size.
+   *
+   * The greater concern is for smaller keys; being able to detect when a
+   * passphrase is say 4 bytes may assist choosing weaker targets. Even though
+   * the loop bounds are constant, we can only actually read key[0..length] so
+   * it doesn't seem possible to make this truly constant time. This should
+   * still hide the length of the key to a leak at the granularity of the cache
+   * line size, which is sufficient for most purposes and similar to the leak
```

Comment:
NIT: I think this assertion about the leak being at the granularity of the cache line is unfounded. Shay Gueron already made a pretty convincing argument that it is not true on x86_64, which he used to justify replacing "small table" AES implementations with ones that with data-invariant memory access patterns. I forgot the details but IIRC at the very least the timing was different when `(uintptr_t)p % CACHE_LINE_WIDTH == 0`.

Reply:
Interesting. I am not sure I've seen that one. There are cache bank conflicts [ala https://eprint.iacr.org/2016/224] but my understanding is those only affect x86 and do not affect the most recent core designs.

My argument implicitly assumed that `key` didn't cross a cache line which is not something we can control, so even without cache bank conflicts it was not valid.